halo2_proofs/src/plonk/vanishing/prover.rs

```diff
@@ -42,10 +49,43 @@ impl<C: CurveAffine> Argument<C> {
         transcript: &mut T,
     ) -> Result<Committed<C>, Error> {
         // Sample a random polynomial of degree n - 1
-        let mut random_poly = domain.empty_coeff();
-        for coeff in random_poly.iter_mut() {
-            *coeff = C::Scalar::random(&mut rng);
-        }
+        #[cfg(feature = "multicore")]
+        let random_poly = {
+            let n_threads = current_num_threads();
+            let n = 1usize << domain.k as usize;
+            let n_chunks = n_threads + if n % n_threads != 0 { 1 } else { 0 };
+            let mut rand_vec = vec![C::Scalar::ZERO; n];
+
+            let mut thread_seeds: Vec<ChaCha20Rng> = (0..n_chunks)
+                .into_iter()
+                .map(|_| {
+                    let mut seed = [0u8; 32];
+                    rng.fill_bytes(&mut seed);
+                    ChaCha20Rng::from_seed(seed)
+                })
+                .collect();
+
+            thread_seeds
+                .par_iter_mut()
+                .zip_eq(rand_vec.par_chunks_mut(n / n_threads))
+                .for_each(|(mut rng, chunk)| {
+                    chunk
+                        .iter_mut()
+                        .for_each(|v| *v = C::Scalar::random(&mut rng))
```

Comment:
```suggestion
                        .for_each(|coeff| *coeff = C::Scalar::random(&mut rng))
```
(to match the sequential code).